crates/common/consensus/lean/src/lib.rs

```diff
@@ -1,4 +1,213 @@
+    || (delta as f64 + 0.25).sqrt() % 1.0 == 0.5 // any x^2+x
+}
+
+// Given a state, output the new state after processing that block
+pub fn process_block(pre_state: &LeanState, block: &Block) -> LeanState {
+    let mut state = pre_state.clone();
+
+    // Track historical blocks in the state
+    // TODO: proper error handlings
+    let _ = state.historical_block_hashes.push(block.parent);
+    let _ = state.justified_slots.push(false);
+
+    while state.historical_block_hashes.len() < block.slot {
+        // TODO: proper error handlings
+        let _ = state.justified_slots.push(false);
+        let _ = state.historical_block_hashes.push(None);
+    }
+
+    // Process votes
+    for vote in &block.votes {
+        // Ignore votes whose source is not already justified,
+        // or whose target is not in the history, or whose target is not a
+        // valid justifiable slot
+        if !state.justified_slots[vote.data.source_slot]
+            || Some(vote.data.source) != state.historical_block_hashes[vote.data.source_slot]
+            || Some(vote.data.target) != state.historical_block_hashes[vote.data.target_slot]
+            || vote.data.target_slot <= vote.data.source_slot
+            || !is_justifiable_slot(&state.latest_finalized_slot, &vote.data.target_slot)
+        {
+            continue;
+        }
+
+        // Track attempts to justify new hashes
+        if !state.justifications.contains_key(&vote.data.target) {
+            let mut empty_justifications = Vec::<bool>::with_capacity(state.num_validators);
+            empty_justifications.resize(state.num_validators, false);
+
+            state
+                .justifications
+                .insert(vote.data.target, empty_justifications);
+        }
+
+        if !state.justifications[&vote.data.target][vote.data.validator_id] {
+            state.justifications.get_mut(&vote.data.target).unwrap()[vote.data.validator_id] = true;
+        }
+
+        let count = state.justifications[&vote.data.target]
+            .iter()
+            .fold(0, |sum, justification| sum + *justification as usize);
+
+        // If 2/3 voted for the same new valid hash to justify
+        if count == (2 * state.num_validators) / 3 {
+            state.latest_justified_hash = vote.data.target;
+            state.latest_justified_slot = vote.data.target_slot;
+            state.justified_slots[vote.data.target_slot] = true;
+
+            state.justifications.remove(&vote.data.target).unwrap();
+
+            // Finalization: if the target is the next valid justifiable
+            // hash after the source
+            let mut is_target_next_valid_justifiable_slot = true;
+
+            for slot in (vote.data.source_slot + 1)..vote.data.target_slot {
+                if is_justifiable_slot(&state.latest_finalized_slot, &slot) {
+                    is_target_next_valid_justifiable_slot = false;
+                    break;
+                }
+            }
```

Comment:
```suggestion
            let is_target_next_valid_justifiable_slot = !((vote.data.source_slot + 1)
                ..vote.data.target_slot)
                .any(|slot| is_justifiable_slot(&state.latest_finalized_slot, &slot));
```

Just my personal preference!